// @receipt-wrangler/receipt-wrangler-core/esm2022/lib/pipes/pipes.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormGetPipe } from './form-get.pipe';
import * as i0 from "@angular/core";
class PipesModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: PipesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.11", ngImport: i0, type: PipesModule, declarations: [FormGetPipe], imports: [CommonModule], exports: [FormGetPipe] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: PipesModule, imports: [CommonModule] }); }
}
export { PipesModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: PipesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [FormGetPipe],
                    imports: [CommonModule],
                    exports: [FormGetPipe],
                }]
        }] });
