// @receipt-wrangler/receipt-wrangler-core/esm2022/lib/base-input/base-input.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { BaseInputComponent } from './base-input/base-input.component';
import * as i0 from "@angular/core";
class BaseInputModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: BaseInputModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.11", ngImport: i0, type: BaseInputModule, declarations: [BaseInputComponent], imports: [CommonModule], exports: [BaseInputComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: BaseInputModule, imports: [CommonModule] }); }
}
export { BaseInputModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: BaseInputModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [BaseInputComponent],
                    imports: [CommonModule],
                    exports: [BaseInputComponent],
                }]
        }] });
