// @receipt-wrangler/receipt-wrangler-core/esm2022/lib/button/button.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MatBadgeModule } from '@angular/material/badge';
import { MatButtonModule } from '@angular/material/button';
import { MatIconModule } from '@angular/material/icon';
import { MatTooltipModule } from '@angular/material/tooltip';
import { RouterModule } from '@angular/router';
import { ButtonComponent } from './button/button.component';
import * as i0 from "@angular/core";
class ButtonModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: ButtonModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.11", ngImport: i0, type: ButtonModule, declarations: [ButtonComponent], imports: [CommonModule,
            MatBadgeModule,
            MatButtonModule,
            MatIconModule,
            MatTooltipModule,
            RouterModule], exports: [ButtonComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: ButtonModule, imports: [CommonModule,
            MatBadgeModule,
            MatButtonModule,
            MatIconModule,
            MatTooltipModule,
            RouterModule] }); }
}
export { ButtonModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: ButtonModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ButtonComponent],
                    imports: [
                        CommonModule,
                        MatBadgeModule,
                        MatButtonModule,
                        MatIconModule,
                        MatTooltipModule,
                        RouterModule,
                    ],
                    exports: [ButtonComponent],
                }]
        }] });
