// @receipt-wrangler/receipt-wrangler-core/esm2022/lib/guards/auth.guard.mjs
import { Injectable } from '@angular/core';
import { Router, } from '@angular/router';
import { Store } from '@ngxs/store';
import { AuthState } from '../store/auth.state';
import { GroupState } from '../store/group.state';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "@ngxs/store";
class AuthGuard {
    constructor(router, store) {
        this.router = router;
        this.store = store;
    }
    canActivate(route, state) {
        const isLoggedIn = this.store.selectSnapshot(AuthState.isLoggedIn);
        const navigatingToAuth = route.url.toString().includes('auth');
        // if user tries to go to login screens while already logged in
        if (navigatingToAuth && isLoggedIn) {
            this.router.navigate([
                this.store.selectSnapshot(GroupState.dashboardLink),
            ]);
            return false;
        }
        else if (navigatingToAuth && !isLoggedIn) {
            return true;
        }
        if (!isLoggedIn) {
            this.router.navigate(['/auth/login']);
        }
        return isLoggedIn;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: AuthGuard, deps: [{ token: i1.Router }, { token: i2.Store }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: AuthGuard, providedIn: 'root' }); }
}
export { AuthGuard };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: AuthGuard, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: function () { return [{ type: i1.Router }, { type: i2.Store }]; } });
