// @receipt-wrangler/receipt-wrangler-core/esm2022/lib/guards/feature.guard.mjs
import { Injectable } from '@angular/core';
import { Store } from '@ngxs/store';
import { FeatureConfigState } from '../store/feature-config.state';
import * as i0 from "@angular/core";
import * as i1 from "@ngxs/store";
class FeatureGuard {
    constructor(store) {
        this.store = store;
    }
    canActivate(route, state) {
        return this.store.selectSnapshot(FeatureConfigState.hasFeature(route.data['feature']));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: FeatureGuard, deps: [{ token: i1.Store }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: FeatureGuard, providedIn: 'root' }); }
}
export { FeatureGuard };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: FeatureGuard, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: function () { return [{ type: i1.Store }]; } });
