// @receipt-wrangler/receipt-wrangler-core/esm2022/lib/base-input/base-input/base-input.component.mjs
import { Component, Input } from '@angular/core';
import { FormControl } from '@angular/forms';
import { map, startWith } from 'rxjs';
import * as i0 from "@angular/core";
class BaseInputComponent {
    constructor() {
        this.inputFormControl = new FormControl();
        this.label = '';
        this.readonly = false;
        this.errorMessages = {};
    }
    ngOnInit() {
        this.errorMessages = {
            required: `${this.label} is required.`,
            email: `${this.label} must be a valid email address.`,
            duplicate: `${this.label} must be unique.`,
            min: `Value must be larger than 0`,
        };
        this.formControlErrors = this.inputFormControl.statusChanges.pipe(startWith(this.inputFormControl.status), map(() => {
            const errors = this.inputFormControl.errors;
            if (errors) {
                const keys = Object.keys(this.inputFormControl.errors);
                return keys.map((k) => {
                    const value = errors[k];
                    let message = '';
                    if (typeof value === 'string') {
                        message = value;
                    }
                    else if (this.errorMessages[k]) {
                        message = this.errorMessages[k];
                    }
                    return {
                        error: k,
                        message: message,
                    };
                });
            }
            else {
                return [];
            }
        }));
        if (this.additionalErrorMessages) {
            this.errorMessages = {
                ...this.errorMessages,
                ...this.additionalErrorMessages,
            };
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: BaseInputComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.11", type: BaseInputComponent, selector: "app-base-input", inputs: { inputFormControl: "inputFormControl", label: "label", additionalErrorMessages: "additionalErrorMessages", readonly: "readonly", placeholder: "placeholder" }, ngImport: i0, template: "<p>base-input works!</p>\n", styles: [""] }); }
}
export { BaseInputComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: BaseInputComponent, decorators: [{
            type: Component,
            args: [{ selector: 'app-base-input', template: "<p>base-input works!</p>\n" }]
        }], propDecorators: { inputFormControl: [{
                type: Input
            }], label: [{
                type: Input
            }], additionalErrorMessages: [{
                type: Input
            }], readonly: [{
                type: Input
            }], placeholder: [{
                type: Input
            }] } });
