// @receipt-wrangler/receipt-wrangler-core/esm2022/lib/directives/directives.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RoleDirective } from './role.directive';
import { FeatureDirective } from './feature.directive';
import * as i0 from "@angular/core";
class DirectivesModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: DirectivesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.11", ngImport: i0, type: DirectivesModule, declarations: [RoleDirective, FeatureDirective], imports: [CommonModule], exports: [RoleDirective, FeatureDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: DirectivesModule, imports: [CommonModule] }); }
}
export { DirectivesModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: DirectivesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [RoleDirective, FeatureDirective],
                    imports: [CommonModule],
                    exports: [RoleDirective, FeatureDirective],
                }]
        }] });
