// @receipt-wrangler/receipt-wrangler-core/esm2022/lib/directives/feature.directive.mjs
import { Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
import { Store } from '@ngxs/store';
import { FeatureConfigState } from '../store/feature-config.state';
import * as i0 from "@angular/core";
import * as i1 from "@ngxs/store";
/**
 * Add the template content to the DOM unless the condition is true.
 */
class FeatureDirective {
    constructor(templateRef, viewContainer, store) {
        this.templateRef = templateRef;
        this.viewContainer = viewContainer;
        this.store = store;
        this.hasView = false;
    }
    set appFeature(feature) {
        const hasFeature = this.store.selectSnapshot(FeatureConfigState.hasFeature(feature));
        if (hasFeature) {
            this.viewContainer.createEmbeddedView(this.templateRef);
            this.hasView = true;
        }
        else if (!hasFeature) {
            this.viewContainer.clear();
            this.hasView = false;
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: FeatureDirective, deps: [{ token: i0.TemplateRef }, { token: i0.ViewContainerRef }, { token: i1.Store }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.11", type: FeatureDirective, selector: "[appFeature]", inputs: { appFeature: "appFeature" }, ngImport: i0 }); }
}
export { FeatureDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: FeatureDirective, decorators: [{
            type: Directive,
            args: [{ selector: '[appFeature]' }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef }, { type: i0.ViewContainerRef }, { type: i1.Store }]; }, propDecorators: { appFeature: [{
                type: Input
            }] } });
