// @receipt-wrangler/receipt-wrangler-core/esm2022/lib/directives/role.directive.mjs
import { Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
import { Store } from '@ngxs/store';
import { AuthState } from '../store/auth.state';
import * as i0 from "@angular/core";
import * as i1 from "@ngxs/store";
class RoleDirective {
    constructor(templateRef, viewContainer, store) {
        this.templateRef = templateRef;
        this.viewContainer = viewContainer;
        this.store = store;
        this.hasView = false;
    }
    set appRole(role) {
        const hasRole = this.store.selectSnapshot(AuthState.hasRole(role));
        if (hasRole) {
            this.viewContainer.createEmbeddedView(this.templateRef);
            this.hasView = true;
        }
        else if (!hasRole) {
            this.viewContainer.clear();
            this.hasView = false;
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: RoleDirective, deps: [{ token: i0.TemplateRef }, { token: i0.ViewContainerRef }, { token: i1.Store }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.11", type: RoleDirective, selector: "[appRole]", inputs: { appRole: "appRole" }, ngImport: i0 }); }
}
export { RoleDirective };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: RoleDirective, decorators: [{
            type: Directive,
            args: [{ selector: '[appRole]' }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef }, { type: i0.ViewContainerRef }, { type: i1.Store }]; }, propDecorators: { appRole: [{
                type: Input
            }] } });
