// @receipt-wrangler/receipt-wrangler-core/esm2022/lib/store/auth.state.mjs
var AuthState_1;
import { __decorate, __metadata } from "tslib";
import { Injectable } from '@angular/core';
import { Action, createSelector, Selector, State, } from '@ngxs/store';
import { User } from '../api/model/user';
import { Logout, SetAuthState, SetUserPreferences } from './auth.state.actions';
import * as i0 from "@angular/core";
let AuthState = AuthState_1 = class AuthState {
    static userPreferences(state) {
        return state.userPreferences;
    }
    static userRole(state) {
        return state.userRole ?? '';
    }
    static isLoggedIn(state) {
        return !AuthState_1.isTokenExpired(state);
    }
    static userId(state) {
        return state.userId ?? '';
    }
    static isTokenExpired(state) {
        if (state.expirationDate) {
            return new Date() >= new Date(Number(state.expirationDate) * 1000);
        }
        else {
            return true;
        }
    }
    static loggedInUser(state) {
        return {
            defaultAvatarColor: state.defaultAvatarColor ?? '',
            displayName: state.displayname ?? '',
            id: Number(state.userId) ?? '',
            username: state.username ?? '',
        };
    }
    static hasRole(role) {
        return createSelector([AuthState_1], (state) => {
            return state.userRole === role;
        });
    }
    setAuthState({ getState, patchState }, payload) {
        const claims = payload.userClaims;
        patchState({
            defaultAvatarColor: claims['DefaultAvatarColor'],
            displayname: claims['Displayname'],
            expirationDate: claims['exp']?.toString(),
            userId: claims['UserId']?.toString(),
            username: claims['Username'],
            userRole: claims['UserRole'],
        });
    }
    logout({ getState, patchState }) {
        patchState({
            defaultAvatarColor: '',
            displayname: '',
            expirationDate: '',
            userId: '',
            username: '',
            userRole: undefined,
            userPreferences: undefined,
        });
    }
    setUserPreferences({ patchState }, payload) {
        patchState({
            userPreferences: payload.userPreferences,
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: AuthState, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: AuthState }); }
};
__decorate([
    Action(SetAuthState),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, SetAuthState]),
    __metadata("design:returntype", void 0)
], AuthState.prototype, "setAuthState", null);
__decorate([
    Action(Logout),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", void 0)
], AuthState.prototype, "logout", null);
__decorate([
    Action(SetUserPreferences),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, SetUserPreferences]),
    __metadata("design:returntype", void 0)
], AuthState.prototype, "setUserPreferences", null);
__decorate([
    Selector(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Object)
], AuthState, "userPreferences", null);
__decorate([
    Selector(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", String)
], AuthState, "userRole", null);
__decorate([
    Selector(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Boolean)
], AuthState, "isLoggedIn", null);
__decorate([
    Selector(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", String)
], AuthState, "userId", null);
__decorate([
    Selector(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Boolean)
], AuthState, "isTokenExpired", null);
__decorate([
    Selector(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Object)
], AuthState, "loggedInUser", null);
AuthState = AuthState_1 = __decorate([
    State({
        name: 'auth',
        defaults: {},
    })
], AuthState);
export { AuthState };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: AuthState, decorators: [{
            type: Injectable
        }], propDecorators: { setAuthState: [], logout: [], setUserPreferences: [] } });
