// @receipt-wrangler/receipt-wrangler-core/esm2022/lib/store/feature-config.state.mjs
var FeatureConfigState_1;
import { __decorate, __metadata } from "tslib";
import { Injectable } from '@angular/core';
import { Action, createSelector, Selector, State, } from '@ngxs/store';
import { SetFeatureConfig } from './feature-config.state.actions';
import * as i0 from "@angular/core";
let FeatureConfigState = FeatureConfigState_1 = class FeatureConfigState {
    static enableLocalSignUp(state) {
        return state.enableLocalSignUp;
    }
    static aiPoweredReceipts(state) {
        return state.aiPoweredReceipts;
    }
    static hasFeature(feature) {
        return createSelector([FeatureConfigState_1], (state) => {
            return !!state[feature];
        });
    }
    setFeatureConfig({ patchState }, payload) {
        patchState({
            aiPoweredReceipts: payload.config?.aiPoweredReceipts,
            enableLocalSignUp: payload.config?.enableLocalSignUp,
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: FeatureConfigState, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: FeatureConfigState }); }
};
__decorate([
    Action(SetFeatureConfig),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, SetFeatureConfig]),
    __metadata("design:returntype", void 0)
], FeatureConfigState.prototype, "setFeatureConfig", null);
__decorate([
    Selector(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Boolean)
], FeatureConfigState, "enableLocalSignUp", null);
__decorate([
    Selector(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Boolean)
], FeatureConfigState, "aiPoweredReceipts", null);
FeatureConfigState = FeatureConfigState_1 = __decorate([
    State({
        name: 'featureConfig',
        defaults: { enableLocalSignUp: true, aiPoweredReceipts: false },
    })
], FeatureConfigState);
export { FeatureConfigState };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: FeatureConfigState, decorators: [{
            type: Injectable
        }], propDecorators: { setFeatureConfig: [] } });
