// @receipt-wrangler/receipt-wrangler-core/esm2022/lib/store/group.state.actions.mjs
class AddGroup {
    static { this.type = '[Group] Add Group'; }
    constructor(group) {
        this.group = group;
    }
}
export { AddGroup };
class RemoveGroup {
    static { this.type = '[Group] Remove Group'; }
    constructor(groupId) {
        this.groupId = groupId;
    }
}
export { RemoveGroup };
class SetGroups {
    static { this.type = '[Group] Set Groups'; }
    constructor(groups) {
        this.groups = groups;
    }
}
export { SetGroups };
class UpdateGroup {
    static { this.type = '[Group] Update Group'; }
    constructor(group) {
        this.group = group;
    }
}
export { UpdateGroup };
class SetSelectedDashboardId {
    static { this.type = '[Group] Set Selected Dashboard Id'; }
    constructor(dashboardId) {
        this.dashboardId = dashboardId;
    }
}
export { SetSelectedDashboardId };
class SetSelectedGroupId {
    static { this.type = '[Group] Set Selected Group Id'; }
    constructor(groupId) {
        this.groupId = groupId;
    }
}
export { SetSelectedGroupId };
