// @receipt-wrangler/receipt-wrangler-core/esm2022/lib/store/user.state.mjs
var UserState_1;
import { __decorate, __metadata } from "tslib";
import { Injectable } from '@angular/core';
import { Action, createSelector, Selector, State, } from '@ngxs/store';
import { AddUser, RemoveUser, SetUsers, UpdateUser, } from './user.state.actions';
import * as i0 from "@angular/core";
let UserState = UserState_1 = class UserState {
    static users(state) {
        return state.users;
    }
    static getUserById(userId) {
        return createSelector([UserState_1], (state) => {
            return state.users.find((u) => u.id.toString() === userId.toString());
        });
    }
    static findUserById(userId) {
        return createSelector([UserState_1], (state) => {
            return state.users.find((u) => u.id.toString() === userId.toString());
        });
    }
    static findUserIndexById(userId, users) {
        return users.findIndex((u) => u.id.toString() === userId);
    }
    setUsers({ getState, patchState }, payload) {
        patchState({
            users: payload.users,
        });
    }
    updateUser({ getState, patchState }, payload) {
        const users = Array.from(getState().users);
        const index = UserState_1.findUserIndexById(payload.userId, users);
        if (index >= 0) {
            users.splice(index, 1, payload.user);
            patchState({
                users: users,
            });
        }
    }
    addUser({ getState, patchState }, payload) {
        const users = Array.from(getState().users);
        users.push(payload.user);
        patchState({
            users: users,
        });
    }
    removeUser({ getState, patchState }, payload) {
        const users = Array.from(getState().users);
        patchState({
            users: users.filter((u) => u.id.toString() !== payload.userId.toString()),
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: UserState, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: UserState }); }
};
__decorate([
    Action(SetUsers),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, SetUsers]),
    __metadata("design:returntype", void 0)
], UserState.prototype, "setUsers", null);
__decorate([
    Action(UpdateUser),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, UpdateUser]),
    __metadata("design:returntype", void 0)
], UserState.prototype, "updateUser", null);
__decorate([
    Action(AddUser),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, AddUser]),
    __metadata("design:returntype", void 0)
], UserState.prototype, "addUser", null);
__decorate([
    Action(RemoveUser),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, RemoveUser]),
    __metadata("design:returntype", void 0)
], UserState.prototype, "removeUser", null);
__decorate([
    Selector(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Array)
], UserState, "users", null);
UserState = UserState_1 = __decorate([
    State({
        name: 'users',
        defaults: {
            users: [],
        },
    })
], UserState);
export { UserState };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: UserState, decorators: [{
            type: Injectable
        }], propDecorators: { setUsers: [], updateUser: [], addUser: [], removeUser: [] } });
