// @receipt-wrangler/receipt-wrangler-core/esm2022/lib/auth/auth.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { ButtonModule } from '../button/button.module';
import { DirectivesModule } from '../directives/directives.module';
import { InputModule } from '../input/input.module';
import { PipesModule } from '../pipes/pipes.module';
import { AuthRoutingModule } from './auth-routing.module';
import { AuthForm } from './sign-up/auth-form.component';
import * as i0 from "@angular/core";
class AuthModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: AuthModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.11", ngImport: i0, type: AuthModule, declarations: [AuthForm], imports: [AuthRoutingModule,
            ButtonModule,
            CommonModule,
            DirectivesModule,
            InputModule,
            PipesModule,
            ReactiveFormsModule], exports: [AuthForm] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: AuthModule, imports: [AuthRoutingModule,
            ButtonModule,
            CommonModule,
            DirectivesModule,
            InputModule,
            PipesModule,
            ReactiveFormsModule] }); }
}
export { AuthModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: AuthModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [AuthForm],
                    imports: [
                        AuthRoutingModule,
                        ButtonModule,
                        CommonModule,
                        DirectivesModule,
                        InputModule,
                        PipesModule,
                        ReactiveFormsModule,
                    ],
                    exports: [AuthForm],
                }]
        }] });
