// @receipt-wrangler/receipt-wrangler-core/esm2022/lib/auth/auth-routing.module.mjs
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { FeatureGuard } from '../guards/feature.guard';
import { AuthForm } from './sign-up/auth-form.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
export const authRoutes = [
    {
        path: 'sign-up',
        component: AuthForm,
        data: {
            isSignUp: true,
            feature: 'enableLocalSignUp',
        },
        canActivate: [FeatureGuard],
    },
    {
        path: 'login',
        component: AuthForm,
    },
];
class AuthRoutingModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: AuthRoutingModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.11", ngImport: i0, type: AuthRoutingModule, imports: [i1.RouterModule], exports: [RouterModule] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: AuthRoutingModule, imports: [RouterModule.forChild(authRoutes), RouterModule] }); }
}
export { AuthRoutingModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: AuthRoutingModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [RouterModule.forChild(authRoutes)],
                    exports: [RouterModule],
                }]
        }] });
