// @receipt-wrangler/receipt-wrangler-core/esm2022/lib/services/app-init.service.mjs
import { Injectable } from '@angular/core';
import { Store } from '@ngxs/store';
import { catchError, forkJoin, switchMap, take, tap, } from 'rxjs';
import { AuthService, FeatureConfigService, GroupsService, UserPreferencesService, UserService, } from '../api';
import { SetFeatureConfig } from '../store/feature-config.state.actions';
import { GroupState } from '../store/group.state';
import { SetGroups, SetSelectedGroupId } from '../store/group.state.actions';
import { SetUsers } from '../store/user.state.actions';
import { ClaimsService } from './claims.service';
import { SetUserPreferences } from '../store';
import * as i0 from "@angular/core";
import * as i1 from "../api";
import * as i2 from "./claims.service";
import * as i3 from "@ngxs/store";
class AppInitService {
    constructor(authService, claimsService, featureConfigService, groupsService, store, userService, userPreferencesService) {
        this.authService = authService;
        this.claimsService = claimsService;
        this.featureConfigService = featureConfigService;
        this.groupsService = groupsService;
        this.store = store;
        this.userService = userService;
        this.userPreferencesService = userPreferencesService;
    }
    initAppData() {
        return new Promise((resolve) => {
            this.featureConfigService
                .getFeatureConfig()
                .pipe(take(1), switchMap((config) => this.store.dispatch(new SetFeatureConfig(config))), catchError((err) => {
                resolve(false);
                return err;
            }), switchMap(() => this.authService.getNewRefreshToken()), switchMap(() => this.getAppData()), tap(() => resolve(true)))
                .subscribe();
        });
    }
    getAppData() {
        const usersCall = this.userService.getUsers().pipe(take(1), tap((users) => this.store.dispatch(new SetUsers(users))));
        const groupsCall = this.groupsService.getGroupsForuser().pipe(take(1), tap((groups) => {
            this.store.dispatch(new SetGroups(groups));
            const groupId = this.store.selectSnapshot(GroupState.selectedGroupId);
            if (!groupId) {
                this.store.dispatch(new SetSelectedGroupId());
            }
        }));
        const userClaims = this.claimsService.getAndSetClaimsForLoggedInUser();
        const userPreferencesCall = this.userPreferencesService
            .getUserPreferences()
            .pipe(take(1), tap((userPreferences) => {
            this.store.dispatch(new SetUserPreferences(userPreferences));
        }));
        return forkJoin(usersCall, groupsCall, userClaims, userPreferencesCall);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: AppInitService, deps: [{ token: i1.AuthService }, { token: i2.ClaimsService }, { token: i1.FeatureConfigService }, { token: i1.GroupsService }, { token: i3.Store }, { token: i1.UserService }, { token: i1.UserPreferencesService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: AppInitService, providedIn: 'root' }); }
}
export { AppInitService };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: AppInitService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: function () { return [{ type: i1.AuthService }, { type: i2.ClaimsService }, { type: i1.FeatureConfigService }, { type: i1.GroupsService }, { type: i3.Store }, { type: i1.UserService }, { type: i1.UserPreferencesService }]; } });
export function initAppData(appInitService) {
    return () => appInitService.initAppData();
}
