// @receipt-wrangler/receipt-wrangler-core/esm2022/lib/services/claims.service.mjs
import { Injectable } from '@angular/core';
import { Store } from '@ngxs/store';
import { switchMap, take } from 'rxjs';
import { UserService } from '../api/api/user.service';
import { SetAuthState } from '../store/auth.state.actions';
import * as i0 from "@angular/core";
import * as i1 from "@ngxs/store";
import * as i2 from "../api/api/user.service";
class ClaimsService {
    constructor(store, userService) {
        this.store = store;
        this.userService = userService;
    }
    getAndSetClaimsForLoggedInUser() {
        return this.userService.getUserClaims().pipe(take(1), switchMap((claims) => this.store.dispatch(new SetAuthState(claims))));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: ClaimsService, deps: [{ token: i1.Store }, { token: i2.UserService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: ClaimsService, providedIn: 'root' }); }
}
export { ClaimsService };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: ClaimsService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: function () { return [{ type: i1.Store }, { type: i2.UserService }]; } });
