// @receipt-wrangler/receipt-wrangler-core/esm2022/lib/services/snackbar.service.mjs
import { Injectable } from '@angular/core';
import { MatSnackBar, } from '@angular/material/snack-bar';
import { DEFAULT_SNACKBAR_ACTION, DEFAULT_SNACKBAR_CONFIG, } from '../constants/snackbar.constant';
import * as i0 from "@angular/core";
import * as i1 from "@angular/material/snack-bar";
class SnackbarService {
    constructor(snackbar) {
        this.snackbar = snackbar;
    }
    error(message) {
        this.snackbar.open(message, DEFAULT_SNACKBAR_ACTION, {
            ...DEFAULT_SNACKBAR_CONFIG,
            panelClass: ['error-snackbar'],
        });
    }
    success(message, configOverrides) {
        this.snackbar.open(message, DEFAULT_SNACKBAR_ACTION, {
            ...DEFAULT_SNACKBAR_CONFIG,
            ...configOverrides,
            panelClass: ['success-snackbar'],
        });
    }
    successFromTemplate(template, configOverrides) {
        return this.snackbar.openFromTemplate(template, {
            ...DEFAULT_SNACKBAR_CONFIG,
            ...configOverrides,
            panelClass: ['success-snackbar'],
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: SnackbarService, deps: [{ token: i1.MatSnackBar }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: SnackbarService, providedIn: 'root' }); }
}
export { SnackbarService };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: SnackbarService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: function () { return [{ type: i1.MatSnackBar }]; } });
