// @receipt-wrangler/receipt-wrangler-core/esm2022/lib/auth/sign-up/auth-form.util.mjs
import { Injectable } from '@angular/core';
import { catchError, map, of, switchMap, tap } from 'rxjs';
import { AuthService } from '../../api';
import { AppInitService, SnackbarService } from '../../services';
import * as i0 from "@angular/core";
import * as i1 from "../../api";
import * as i2 from "../../services";
class AuthFormUtil {
    constructor(authService, snackbarService, appInitService) {
        this.authService = authService;
        this.snackbarService = snackbarService;
        this.appInitService = appInitService;
    }
    getSubmitObservable(form, isSignUp) {
        const isValid = form.valid;
        if (isValid && isSignUp) {
            return this.authService.signUp(form.value).pipe(tap(() => {
                this.snackbarService.success('User successfully signed up');
            }), catchError((err) => of(this.snackbarService.error(err.error['username'] ?? err['errMsg']))));
        }
        else if (isValid && !isSignUp) {
            return this.authService.login(form.value).pipe(tap(() => {
                this.snackbarService.success('Successfully logged in');
            }), switchMap(() => this.appInitService.getAppData()), map(() => undefined));
        }
        else {
            return of(undefined);
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: AuthFormUtil, deps: [{ token: i1.AuthService }, { token: i2.SnackbarService }, { token: i2.AppInitService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: AuthFormUtil }); }
}
export { AuthFormUtil };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: AuthFormUtil, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.AuthService }, { type: i2.SnackbarService }, { type: i2.AppInitService }]; } });
