// @receipt-wrangler/receipt-wrangler-core/esm2022/src/lib/api/api.module.mjs
import { NgModule, SkipSelf, Optional } from '@angular/core';
import { Configuration } from './configuration';
import { HttpClient } from '@angular/common/http';
import { AuthService } from './api/auth.service';
import { CategoryService } from './api/category.service';
import { CommentService } from './api/comment.service';
import { DashboardService } from './api/dashboard.service';
import { FeatureConfigService } from './api/featureConfig.service';
import { GroupsService } from './api/groups.service';
import { NotificationsService } from './api/notifications.service';
import { ReceiptService } from './api/receipt.service';
import { ReceiptImageService } from './api/receiptImage.service';
import { SearchService } from './api/search.service';
import { TagService } from './api/tag.service';
import { UserService } from './api/user.service';
import { UserPreferencesService } from './api/userPreferences.service';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
class ApiModule {
    static forRoot(configurationFactory) {
        return {
            ngModule: ApiModule,
            providers: [{ provide: Configuration, useFactory: configurationFactory }]
        };
    }
    constructor(parentModule, http) {
        if (parentModule) {
            throw new Error('ApiModule is already loaded. Import in your base AppModule only.');
        }
        if (!http) {
            throw new Error('You need to import the HttpClientModule in your AppModule! \n' +
                'See also https://github.com/angular/angular/issues/20575');
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: ApiModule, deps: [{ token: ApiModule, optional: true, skipSelf: true }, { token: i1.HttpClient, optional: true }], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.11", ngImport: i0, type: ApiModule }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: ApiModule, providers: [
            AuthService,
            CategoryService,
            CommentService,
            DashboardService,
            FeatureConfigService,
            GroupsService,
            NotificationsService,
            ReceiptService,
            ReceiptImageService,
            SearchService,
            TagService,
            UserService,
            UserPreferencesService
        ] }); }
}
export { ApiModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.11", ngImport: i0, type: ApiModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [],
                    declarations: [],
                    exports: [],
                    providers: [
                        AuthService,
                        CategoryService,
                        CommentService,
                        DashboardService,
                        FeatureConfigService,
                        GroupsService,
                        NotificationsService,
                        ReceiptService,
                        ReceiptImageService,
                        SearchService,
                        TagService,
                        UserService,
                        UserPreferencesService
                    ]
                }]
        }], ctorParameters: function () { return [{ type: ApiModule, decorators: [{
                    type: Optional
                }, {
                    type: SkipSelf
                }] }, { type: i1.HttpClient, decorators: [{
                    type: Optional
                }] }]; } });
