// @angular-eslint/eslint-plugin/dist/rules/prefer-output-readonly.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RULE_NAME = void 0;
const utils_1 = require("@angular-eslint/utils");
const create_eslint_rule_1 = require("../utils/create-eslint-rule");
exports.RULE_NAME = 'prefer-output-readonly';
exports.default = (0, create_eslint_rule_1.createESLintRule)({
    name: exports.RULE_NAME,
    meta: {
        type: 'suggestion',
        docs: {
            description: 'Prefer to declare `@Output` as `readonly` since they are not supposed to be reassigned',
            recommended: false,
        },
        hasSuggestions: true,
        schema: [],
        messages: {
            preferOutputReadonly: 'Prefer to declare `@Output` as `readonly` since they are not supposed to be reassigned',
            suggestAddReadonlyModifier: 'Add `readonly` modifier',
        },
    },
    defaultOptions: [],
    create(context) {
        return {
            [`PropertyDefinition:not([readonly]) > ${utils_1.Selectors.OUTPUT_DECORATOR}`]({ parent: { key }, }) {
                context.report({
                    node: key,
                    messageId: 'preferOutputReadonly',
                    suggest: [
                        {
                            messageId: 'suggestAddReadonlyModifier',
                            fix: (fixer) => fixer.insertTextBefore(key, 'readonly '),
                        },
                    ],
                });
            },
        };
    },
});
